src/Controller.js

```diff
@@ -0,0 +1,123 @@
+import FunctionTree from 'function-tree'
+import Module from './Module'
+import DefaultModel from './DefaultModel'
+import {ensurePath, isDeveloping, throwError} from './utils'
+import VerifyInputProvider from './providers/VerifyInput'
+import ContextProvider from 'function-tree/providers/Context'
+import StateProvider from './providers/State'
+import DebuggerProvider from './providers/Debugger'
+import ControllerProvider from './providers/Controller'
+import {EventEmitter} from 'events'
+import {dependencyStore as computedDependencyStore} from './Computed'
+
+/*
+  The controller is where everything is attached. The devtools
+  and router is attached directly. Also a top level module is created.
+  The controller creates the function tree that will run all signals,
+  based on top level providers and providers defined in modules
+*/
+class Controller extends EventEmitter {
+  constructor({state = {}, routes = {}, signals = {}, providers = [], modules = {}, router, devtools}) {
+    super()
+    this.flush = this.flush.bind(this)
+    this.devtools = devtools
+    this.model = new DefaultModel({})
+    this.module = new Module([], () => ({
+      state,
+      routes,
+      signals,
+      modules
+    }), {controller: this})
+    this.router = router ? router(this.module.getRoutes(), this) : null
+
+    const allProviders = (
+      this.router ? [
+        this.router.provider
+      ] : []
+    ).concat((
+      this.devtools ? [
+        DebuggerProvider(this.devtools)
+      ] : []
+    )).concat((
+      isDeveloping() ? [
+        VerifyInputProvider
+      ] : []
+    )).concat(
+      ControllerProvider(this),
+      StateProvider(this.model)
+    ).concat(
+      providers.concat(this.module.getProviders())
+        .map(provider => typeof provider === 'function' ? provider : ContextProvider(provider))
+    )
+
+    this.runTree = new FunctionTree(allProviders)
+    this.runTree.on('asyncFunction', this.flush)
+    this.runTree.on('end', this.flush)
+
+    if (this.devtools) {
```

Comment:
Hehe, I see your point. The thing is that we try to squeeze them into the "community modules" circle, when they in fact are core features of Cerebral... which are square... so, does not fit... kinda ;-)

For example the router needs to expose a method to components (convert signal to url). Not a community module requirement. The router now also extracts "routes" properties from all modules. Not a community module requirement.

Also squeezing these core features into the concept of a "community module" makes them behave differently, because they are special. For example the devtools has to be the first provider registered to correctly trigger a signal execution to the debugger, which means if it was a "community module" it would always have to be put first... which is a gotcha. Been at war with Redux because of ordering, causing so many weird issues.

What I mean to say is that the router and devtools are special core features that should be integrated to keep the public API as clean and simple as possible. Tweaking the public API to enable what these features need is not good I think.

Also what we are seeing now is that modules will not be such a common thing. `forms` will just be a set of helpers to define state and compose chains. `firebase` will become just a provider. `fuse` is a computed. `http` is overkill... reinventing the wheel. Now with `function-tree` you can just attach axios/request or whatever as a provider and you get the debugging experience and all out of the box.

On the new website I want to remove `community-modules` and just call it "Community tools" or whatever. They can be computeds (fuse), modules (useragent), providers (firebase), collection of factories and actions (forms), boilerplates etc.